### PR TITLE
V2 Long Haul test run report

### PR DESCRIPTION
**Pull request by tmahmood-microsoft**

### CODE REVIEW

File: e2e/LongHaul/LongHaul report/LongHaul Report.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
+# Long haul test results
+
+This document describes long haul test results for Azure IoT SDK for C# v2. The results are based on 4 IoT Hub devices (MQTT/TCP, MQTT/WS, AMQP/TCP, AMQP/WS) and 2 IoT Edge modules (MQTT/TCP, AMQP/TCP). These tests ran for 64 hours.
```
Comment (drwill-ms) [inline e2e/LongHaul/LongHaul report/LongHaul Report.md:3]:
```suggestion
This document describes long haul test results for Azure IoT SDK for C# v2. The results are based on 4 IoT Hub devices (MQTT/TCP, MQTT/WS, AMQP/TCP, AMQP/WS) and 2 IoT Edge modules (MQTT/TCP, AMQP/TCP). These tests ran for 64 hours and were stopped manually.
```

File: e2e/LongHaul/LongHaul report/LongHaul Report.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+These long haul tests are based on executing different operations like c2d messages, direct method, telemetry messages, feedback messages and file uploads to service client. Each of these operations is executed over regular intervals and logs the accumulated total count of the operations and the time it takes for successful completion. The graphs used here are results of Kusto queries executed over these collected logs.
```
Comment (drwill-ms) [inline e2e/LongHaul/LongHaul report/LongHaul Report.md:4]:
```suggestion
These long haul tests are based on executing different operations like C2D messages, direct method, telemetry messages, feedback messages and file uploads to service client. Each of these operations is executed over regular intervals and logs the accumulated total count of the operations and the time it takes for successful completion. The graphs used here are results of Kusto queries executed over these collected logs.
```

File: e2e/LongHaul/LongHaul report/LongHaul Report.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+### Garbage collector available bytes
+
+The tests keep track of total available garbage collector bytes. This allows for detecting any memory leaks in the program. No significant memory leaks were observed during the tests. The screenshot below shows data from garbage collector over all the 4 devices used in the tests.
+The data shows consistent size of available garbage collector bytes.
+
+![_](GCbytes.png)
+
+### Desired twin and C2D messages handled
+
+Each device in the test receives desired twin properties updates and C2D messages every 3 seconds from service. The screenshot below shows data collected over all 4 devices used in the tests. The graph is not a linear curve due a bug in MQTT implementation where desired twin properties and c2d messages were not being parsed correctly. This test helped us find and resolve the issue.
```
Comment (drwill-ms) [inline e2e/LongHaul/LongHaul report/LongHaul Report.md:23]:
```suggestion
Each device in the test receives desired twin properties updates and C2D messages every 3 seconds from a service app. The screenshot below shows data collected over all 4 devices used in the tests. The graph is not a linear curve due a bug in MQTT implementation where desired twin properties and c2d messages were not being parsed correctly. This test helped us find and resolve the issue.
```

File: e2e/LongHaul/LongHaul report/LongHaul Report.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+![_](c2d,twin.png)
+
+### Reported twin and Telemetry sent
```
Comment (drwill-ms) [inline e2e/LongHaul/LongHaul report/LongHaul Report.md:27]: nit: consistency of casing of titles

```suggestion
### Reported twin and telemetry sent
```

File: e2e/LongHaul/LongHaul report/LongHaul Report.md [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+### Direct method delay
+
+Each device logs the time (seconds) it takes to successfully receive direct method from service. The graph below observes time < 1s to receive direct method with exceptions where a device was reconnecting and the delay was ~5-6 seconds.
+
+![_](DirectMethodDelay.png)
+
+### C2D message delay
+
+Each device logs the time (seconds) it takes to successfully receive c2d message from service. The graph below observes time < 1s to receive c2d message with exceptions where a device was reconnecting and the delay was ~5-6 seconds.
```
Comment (drwill-ms) [inline e2e/LongHaul/LongHaul report/LongHaul Report.md:65]:
```suggestion
Each device logs the time (seconds) it takes to successfully receive C2D message from service. The graph below observes time < 1s to receive C2D message with exceptions where a device was reconnecting, and the delay was ~5-6 seconds.
```

File: e2e/LongHaul/LongHaul report/LongHaul Report.md [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+![_](c2dDelay.png)
+
+
+## IoT Edge module client test results
+
+### Garbage collector available bytes
+
+The tests keep track of total available garbage collector bytes. This allows for detecting any memory leaks in the program. No significant memory leaks were observed during the tests. The screenshot below shows data from garbage collector over 2 edge modules used in the tests.
```
Comment (drwill-ms) [inline e2e/LongHaul/LongHaul report/LongHaul Report.md:74]:
```suggestion
The tests keep track of total available garbage collector bytes. This allows for detecting any memory leaks increasing memory used over time in the program. No significant memory leaks were observed during the tests. The screenshot below shows data from garbage collector over 2 edge modules used in the tests.
```

File: e2e/LongHaul/LongHaul report/LongHaul Report.md [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+![_](moduleGCbytes.png)
+
+### Disconnection Duration
+
+MQTT protocol has a limitation where the device disconnects before token refresh and then reconnects. The device client sets TTL to 60 minutes by default, thus we see a reconnection after every ~55 minutes. The graph below shows the time duration (seconds) for which MQTT devices disconnects. Apart from a single exception, edge module reconnect within 1 second.
+
+![_](ModuleDisconnectionSeconds.png)
+
+### Telemetry Message Backlog
```
Comment (drwill-ms) [inline e2e/LongHaul/LongHaul report/LongHaul Report.md:85]: Isn't this a repeat of what is on line 39?

Comment (drwill-ms) [reply e2e/LongHaul/LongHaul report/LongHaul Report.md:85]: Oh, or a repeat for module client.

**Review by drwill-ms — APPROVED** [review]
Mostly just casing consistency issues and a few minor text corrections.

File: e2e/LongHaul/LongHaul report/LongHaul Report.md (hunk 7 above)
Comment (tmahmood-microsoft) [reply e2e/LongHaul/LongHaul report/LongHaul Report.md:85]: line 39 is for iot hub device,. while this one is for edge module but yes, similar text for both.

**Pull request merged**